// @angular-devkit/core/src/virtual-fs/host/scoped.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = require("../path");
class ScopedHost {
    constructor(_delegate, _root = path_1.NormalizedRoot) {
        this._delegate = _delegate;
        this._root = _root;
    }
    get capabilities() { return this._delegate.capabilities; }
    write(path, content) {
        return this._delegate.write(path_1.join(this._root, path), content);
    }
    read(path) {
        return this._delegate.read(path_1.join(this._root, path));
    }
    delete(path) {
        return this._delegate.delete(path_1.join(this._root, path));
    }
    rename(from, to) {
        return this._delegate.rename(path_1.join(this._root, from), path_1.join(this._root, to));
    }
    list(path) {
        return this._delegate.list(path_1.join(this._root, path));
    }
    exists(path) {
        return this._delegate.exists(path_1.join(this._root, path));
    }
    isDirectory(path) {
        return this._delegate.isDirectory(path_1.join(this._root, path));
    }
    isFile(path) {
        return this._delegate.isFile(path_1.join(this._root, path));
    }
    // Some hosts may not support stats.
    stats(path) {
        return this._delegate.stats(path_1.join(this._root, path));
    }
    // Some hosts may not support watching.
    watch(path, options) {
        return this._delegate.watch(path_1.join(this._root, path), options);
    }
}
exports.ScopedHost = ScopedHost;
