// @angular-devkit/core/src/terminal/colors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
const object_1 = require("../utils/object");
const kColors = {
    modifiers: {
        reset: [0, 0],
        bold: [1, 22],
        dim: [2, 22],
        italic: [3, 23],
        underline: [4, 24],
        inverse: [7, 27],
        hidden: [8, 28],
        strikethrough: [9, 29],
    },
    colors: {
        black: [30, 39],
        red: [31, 39],
        green: [32, 39],
        yellow: [33, 39],
        blue: [34, 39],
        magenta: [35, 39],
        cyan: [36, 39],
        white: [37, 39],
        gray: [90, 39],
    },
    bgColors: {
        bgBlack: [40, 49],
        bgRed: [41, 49],
        bgGreen: [42, 49],
        bgYellow: [43, 49],
        bgBlue: [44, 49],
        bgMagenta: [45, 49],
        bgCyan: [46, 49],
        bgWhite: [47, 49],
    },
};
const kColorFunctions = object_1.mapObject(kColors, (_, v) => {
    return object_1.mapObject(v, (_, vv) => (x) => `\u001b[${vv[0]}m${x}\u001b[${vv[1]}m`);
});
exports.reset = kColorFunctions.modifiers.reset;
exports.bold = kColorFunctions.modifiers.bold;
exports.dim = kColorFunctions.modifiers.dim;
exports.italic = kColorFunctions.modifiers.italic;
exports.underline = kColorFunctions.modifiers.underline;
exports.inverse = kColorFunctions.modifiers.inverse;
exports.hidden = kColorFunctions.modifiers.hidden;
exports.strikethrough = kColorFunctions.modifiers.strikethrough;
exports.black = kColorFunctions.colors.black;
exports.red = kColorFunctions.colors.red;
exports.green = kColorFunctions.colors.green;
exports.yellow = kColorFunctions.colors.yellow;
exports.blue = kColorFunctions.colors.blue;
exports.magenta = kColorFunctions.colors.magenta;
exports.cyan = kColorFunctions.colors.cyan;
exports.white = kColorFunctions.colors.white;
exports.grey = kColorFunctions.colors.gray;
exports.gray = kColorFunctions.colors.gray;
exports.bgBlack = kColorFunctions.bgColors.bgBlack;
exports.bgRed = kColorFunctions.bgColors.bgRed;
exports.bgGreen = kColorFunctions.bgColors.bgGreen;
exports.bgYellow = kColorFunctions.bgColors.bgYellow;
exports.bgBlue = kColorFunctions.bgColors.bgBlue;
exports.bgMagenta = kColorFunctions.bgColors.bgMagenta;
exports.bgCyan = kColorFunctions.bgColors.bgCyan;
exports.bgWhite = kColorFunctions.bgColors.bgWhite;
