// ng2-bs3-modal/bundles/ng2-bs3-modal.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
System.register("ng2-bs3-modal/components/modal-instance", ['rxjs/Observable', 'rxjs/add/operator/map', 'rxjs/add/observable/fromEvent'], function(exports_1, context_1) {
    "use strict";
    var __moduleName = context_1 && context_1.id;
    var Observable_1;
    var ModalInstance, ModalResult;
    function booleanOrValue(value) {
        if (value === 'true')
            return true;
        else if (value === 'false')
            return false;
        return value;
    }
    function toPromise(observable) {
        return new Promise(function (resolve, reject) {
            observable.subscribe(function (next) {
                resolve(next);
            });
        });
    }
    return {
        setters:[
            function (Observable_1_1) {
                Observable_1 = Observable_1_1;
            },
            function (_1) {},
            function (_2) {}],
        execute: function() {
            ModalInstance = (function () {
                function ModalInstance(element) {
                    this.element = element;
                    this.suffix = '.ng2-bs3-modal';
                    this.shownEventName = 'shown.bs.modal' + this.suffix;
                    this.hiddenEventName = 'hidden.bs.modal' + this.suffix;
                    this.visible = false;
                    this.init();
                }
                ModalInstance.prototype.open = function () {
                    return this.show();
                };
                ModalInstance.prototype.close = function () {
                    this.result = ModalResult.Close;
                    return this.hide();
                };
                ModalInstance.prototype.dismiss = function () {
                    this.result = ModalResult.Dismiss;
                    return this.hide();
                };
                ModalInstance.prototype.destroy = function () {
                    var _this = this;
                    return this.hide().then(function () {
                        if (_this.$modal) {
                            _this.$modal.data('bs.modal', null);
                            _this.$modal.remove();
                        }
                    });
                };
                ModalInstance.prototype.show = function () {
                    var promise = toPromise(this.shown);
                    this.resetData();
                    this.$modal.modal();
                    return promise;
                };
                ModalInstance.prototype.hide = function () {
                    if (this.$modal && this.visible) {
                        var promise = toPromise(this.hidden);
                        this.$modal.modal('hide');
                        return promise;
                    }
                    return Promise.resolve(this.result);
                };
                ModalInstance.prototype.init = function () {
                    var _this = this;
                    this.$modal = jQuery(this.element.nativeElement);
                    this.$modal.appendTo('body');
                    this.shown = Observable_1.Observable.fromEvent(this.$modal, this.shownEventName)
                        .map(function () {
                        _this.visible = true;
                    });
                    this.hidden = Observable_1.Observable.fromEvent(this.$modal, this.hiddenEventName)
                        .map(function () {
                        var result = (!_this.result || _this.result === ModalResult.None)
                            ? ModalResult.Dismiss : _this.result;
                        _this.result = ModalResult.None;
                        _this.visible = false;
                        return result;
                    });
                };
                ModalInstance.prototype.resetData = function () {
                    this.$modal.removeData();
                    this.$modal.data('backdrop', booleanOrValue(this.$modal.attr('data-backdrop')));
                    this.$modal.data('keyboard', booleanOrValue(this.$modal.attr('data-keyboard')));
                };
                return ModalInstance;
            }());
            exports_1("ModalInstance", ModalInstance);
            (function (ModalResult) {
                ModalResult[ModalResult["None"] = 0] = "None";
                ModalResult[ModalResult["Close"] = 1] = "Close";
                ModalResult[ModalResult["Dismiss"] = 2] = "Dismiss";
            })(ModalResult || (ModalResult = {}));
            exports_1("ModalResult", ModalResult);
        }
    }
});
System.register("ng2-bs3-modal/components/modal", ['@angular/core', "ng2-bs3-modal/components/modal-instance"], function(exports_2, context_2) {
    "use strict";
    var __moduleName = context_2 && context_2.id;
    var core_1, modal_instance_1;
    var ModalComponent, ModalSize;
    return {
        setters:[
            function (core_1_1) {
                core_1 = core_1_1;
            },
            function (modal_instance_1_1) {
                modal_instance_1 = modal_instance_1_1;
            }],
        execute: function() {
            ModalComponent = (function () {
                function ModalComponent(element) {
                    var _this = this;
                    this.element = element;
                    this.overrideSize = null;
                    this.visible = false;
                    this.animation = true;
                    this.backdrop = true;
                    this.keyboard = true;
                    this.cssClass = '';
                    this.onClose = new core_1.EventEmitter(false);
                    this.onDismiss = new core_1.EventEmitter(false);
                    this.onOpen = new core_1.EventEmitter(false);
                    this.instance = new modal_instance_1.ModalInstance(this.element);
                    this.instance.hidden.subscribe(function (result) {
                        _this.visible = _this.instance.visible;
                        if (result === modal_instance_1.ModalResult.Dismiss) {
                            _this.onDismiss.emit(undefined);
                        }
                    });
                    this.instance.shown.subscribe(function () {
                        _this.onOpen.emit(undefined);
                    });
                }
                Object.defineProperty(ModalComponent.prototype, "fadeClass", {
                    get: function () {
                        return this.animation;
                    },
                    enumerable: true,
                    configurable: true
                });
                Object.defineProperty(ModalComponent.prototype, "dataKeyboardAttr", {
                    get: function () {
                        return this.keyboard;
                    },
                    enumerable: true,
                    configurable: true
                });
                Object.defineProperty(ModalComponent.prototype, "dataBackdropAttr", {
                    get: function () {
                        return this.backdrop;
                    },
                    enumerable: true,
                    configurable: true
                });
                ModalComponent.prototype.ngOnDestroy = function () {
                    return this.instance && this.instance.destroy();
                };
                ModalComponent.prototype.routerCanDeactivate = function () {
                    return this.ngOnDestroy();
                };
                ModalComponent.prototype.open = function (size) {
                    var _this = this;
                    if (ModalSize.validSize(size))
                        this.overrideSize = size;
                    return this.instance.open().then(function () {
                        _this.visible = _this.instance.visible;
                    });
                };
                ModalComponent.prototype.close = function (value) {
                    var _this = this;
                    return this.instance.close().then(function () {
                        _this.onClose.emit(value);
                    });
                };
                ModalComponent.prototype.dismiss = function () {
                    return this.instance.dismiss();
                };
                ModalComponent.prototype.getCssClasses = function () {
                    var classes = [];
                    if (this.isSmall()) {
                        classes.push('modal-sm');
                    }
                    if (this.isLarge()) {
                        classes.push('modal-lg');
                    }
                    if (this.cssClass !== '') {
                        classes.push(this.cssClass);
                    }
                    return classes.join(' ');
                };
                ModalComponent.prototype.isSmall = function () {
                    return this.overrideSize !== ModalSize.Large
                        && this.size === ModalSize.Small
                        || this.overrideSize === ModalSize.Small;
                };
                ModalComponent.prototype.isLarge = function () {
                    return this.overrideSize !== ModalSize.Small
                        && this.size === ModalSize.Large
                        || this.overrideSize === ModalSize.Large;
                };
                __decorate([
                    core_1.Input(), 
                    __metadata('design:type', Boolean)
                ], ModalComponent.prototype, "animation", void 0);
                __decorate([
                    core_1.Input(), 
                    __metadata('design:type', Object)
                ], ModalComponent.prototype, "backdrop", void 0);
                __decorate([
                    core_1.Input(), 
                    __metadata('design:type', Boolean)
                ], ModalComponent.prototype, "keyboard", void 0);
                __decorate([
                    core_1.Input(), 
                    __metadata('design:type', String)
                ], ModalComponent.prototype, "size", void 0);
                __decorate([
                    core_1.Input(), 
                    __metadata('design:type', String)
                ], ModalComponent.prototype, "cssClass", void 0);
                __decorate([
                    core_1.Output(), 
                    __metadata('design:type', core_1.EventEmitter)
                ], ModalComponent.prototype, "onClose", void 0);
                __decorate([
                    core_1.Output(), 
                    __metadata('design:type', core_1.EventEmitter)
                ], ModalComponent.prototype, "onDismiss", void 0);
                __decorate([
                    core_1.Output(), 
                    __metadata('design:type', core_1.EventEmitter)
                ], ModalComponent.prototype, "onOpen", void 0);
                __decorate([
                    core_1.HostBinding('class.fade'), 
                    __metadata('design:type', Boolean)
                ], ModalComponent.prototype, "fadeClass", null);
                __decorate([
                    core_1.HostBinding('attr.data-keyboard'), 
                    __metadata('design:type', Boolean)
                ], ModalComponent.prototype, "dataKeyboardAttr", null);
                __decorate([
                    core_1.HostBinding('attr.data-backdrop'), 
                    __metadata('design:type', Object)
                ], ModalComponent.prototype, "dataBackdropAttr", null);
                ModalComponent = __decorate([
                    core_1.Component({
                        selector: 'modal',
                        host: {
                            'class': 'modal',
                            'role': 'dialog',
                            'tabindex': '-1'
                        },
                        template: "\n        <div class=\"modal-dialog\" [ngClass]=\"getCssClasses()\">\n            <div class=\"modal-content\">\n                <ng-content></ng-content>\n            </div>\n        </div>\n    "
                    }), 
                    __metadata('design:paramtypes', [core_1.ElementRef])
                ], ModalComponent);
                return ModalComponent;
            }());
            exports_2("ModalComponent", ModalComponent);
            ModalSize = (function () {
                function ModalSize() {
                }
                ModalSize.validSize = function (size) {
                    return size && (size === ModalSize.Small || size === ModalSize.Large);
                };
                ModalSize.Small = 'sm';
                ModalSize.Large = 'lg';
                return ModalSize;
            }());
            exports_2("ModalSize", ModalSize);
        }
    }
});
System.register("ng2-bs3-modal/components/modal-body", ['@angular/core'], function(exports_3, context_3) {
    "use strict";
    var __moduleName = context_3 && context_3.id;
    var core_2;
    var ModalBodyComponent;
    return {
        setters:[
            function (core_2_1) {
                core_2 = core_2_1;
            }],
        execute: function() {
            ModalBodyComponent = (function () {
                function ModalBodyComponent() {
                }
                ModalBodyComponent = __decorate([
                    core_2.Component({
                        selector: 'modal-body',
                        template: "\n        <div class=\"modal-body\">\n            <ng-content></ng-content>\n        </div>\n    "
                    }), 
                    __metadata('design:paramtypes', [])
                ], ModalBodyComponent);
                return ModalBodyComponent;
            }());
            exports_3("ModalBodyComponent", ModalBodyComponent);
        }
    }
});
/**
 * This file is generated by the Angular 2 template compiler.
 * Do not edit.
 */
/* tslint:disable */
System.register("ng2-bs3-modal/components/modal-body.ngfactory", ['@angular/core/src/linker/view', '@angular/core/src/linker/element', "ng2-bs3-modal/components/modal-body", '@angular/core/src/linker/view_utils', '@angular/core/src/linker/view_type', '@angular/core/src/change_detection/change_detection', '@angular/core/src/metadata/view', '@angular/core/src/linker/component_factory'], function(exports_4, context_4) {
    "use strict";
    var __moduleName = context_4 && context_4.id;
    var import1, import2, import3, import4, import6, import7, import8, import9;
    var renderType_ModalBodyComponent_Host, _View_ModalBodyComponent_Host0, ModalBodyComponentNgFactory, styles_ModalBodyComponent, renderType_ModalBodyComponent, _View_ModalBodyComponent0;
    function viewFactory_ModalBodyComponent_Host0(viewUtils, parentInjector, declarationEl) {
        if ((renderType_ModalBodyComponent_Host === null)) {
            (renderType_ModalBodyComponent_Host = viewUtils.createRenderComponentType('', 0, import8.ViewEncapsulation.None, [], {}));
        }
        return new _View_ModalBodyComponent_Host0(viewUtils, parentInjector, declarationEl);
    }
    function viewFactory_ModalBodyComponent0(viewUtils, parentInjector, declarationEl) {
        if ((renderType_ModalBodyComponent === null)) {
            (renderType_ModalBodyComponent = viewUtils.createRenderComponentType('/Users/doug/Source/dougludlow/ng2-bs3-modal/src/ng2-bs3-modal/components/modal-body.ts class ModalBodyComponent - inline template', 1, import8.ViewEncapsulation.None, styles_ModalBodyComponent, {}));
        }
        return new _View_ModalBodyComponent0(viewUtils, parentInjector, declarationEl);
    }
    exports_4("viewFactory_ModalBodyComponent0", viewFactory_ModalBodyComponent0);
    return {
        setters:[
            function (import1_1) {
                import1 = import1_1;
            },
            function (import2_1) {
                import2 = import2_1;
            },
            function (import3_1) {
                import3 = import3_1;
            },
            function (import4_1) {
                import4 = import4_1;
            },
            function (import6_1) {
                import6 = import6_1;
            },
            function (import7_1) {
                import7 = import7_1;
            },
            function (import8_1) {
                import8 = import8_1;
            },
            function (import9_1) {
                import9 = import9_1;
            }],
        execute: function() {
            renderType_ModalBodyComponent_Host = null;
            _View_ModalBodyComponent_Host0 = (function (_super) {
                __extends(_View_ModalBodyComponent_Host0, _super);
                function _View_ModalBodyComponent_Host0(viewUtils, parentInjector, declarationEl) {
                    _super.call(this, _View_ModalBodyComponent_Host0, renderType_ModalBodyComponent_Host, import6.ViewType.HOST, viewUtils, parentInjector, declarationEl, import7.ChangeDetectorStatus.CheckAlways);
                }
                _View_ModalBodyComponent_Host0.prototype.createInternal = function (rootSelector) {
                    this._el_0 = this.selectOrCreateHostElement('modal-body', rootSelector, null);
                    this._appEl_0 = new import2.AppElement(0, null, this, this._el_0);
                    var compView_0 = viewFactory_ModalBodyComponent0(this.viewUtils, this.injector(0), this._appEl_0);
                    this._ModalBodyComponent_0_4 = new import3.ModalBodyComponent();
                    this._appEl_0.initComponent(this._ModalBodyComponent_0_4, [], compView_0);
                    compView_0.create(this._ModalBodyComponent_0_4, this.projectableNodes, null);
                    this.init([].concat([this._el_0]), [this._el_0], [], []);
                    return this._appEl_0;
                };
                _View_ModalBodyComponent_Host0.prototype.injectorGetInternal = function (token, requestNodeIndex, notFoundResult) {
                    if (((token === import3.ModalBodyComponent) && (0 === requestNodeIndex))) {
                        return this._ModalBodyComponent_0_4;
                    }
                    return notFoundResult;
                };
                return _View_ModalBodyComponent_Host0;
            }(import1.AppView));
            exports_4("ModalBodyComponentNgFactory", ModalBodyComponentNgFactory = new import9.ComponentFactory('modal-body', viewFactory_ModalBodyComponent_Host0, import3.ModalBodyComponent));
            styles_ModalBodyComponent = [];
            renderType_ModalBodyComponent = null;
            _View_ModalBodyComponent0 = (function (_super) {
                __extends(_View_ModalBodyComponent0, _super);
                function _View_ModalBodyComponent0(viewUtils, parentInjector, declarationEl) {
                    _super.call(this, _View_ModalBodyComponent0, renderType_ModalBodyComponent, import6.ViewType.COMPONENT, viewUtils, parentInjector, declarationEl, import7.ChangeDetectorStatus.CheckAlways);
                }
                _View_ModalBodyComponent0.prototype.createInternal = function (rootSelector) {
                    var parentRenderNode = this.renderer.createViewRoot(this.declarationAppElement.nativeElement);
                    this._text_0 = this.renderer.createText(parentRenderNode, '\n        ', null);
                    this._el_1 = this.renderer.createElement(parentRenderNode, 'div', null);
                    this.renderer.setElementAttribute(this._el_1, 'class', 'modal-body');
                    this._text_2 = this.renderer.createText(this._el_1, '\n            ', null);
                    this.renderer.projectNodes(this._el_1, import4.flattenNestedViewRenderNodes(this.projectableNodes[0]));
                    this._text_3 = this.renderer.createText(this._el_1, '\n        ', null);
                    this._text_4 = this.renderer.createText(parentRenderNode, '\n    ', null);
                    this.init([], [
                        this._text_0,
                        this._el_1,
                        this._text_2,
                        this._text_3,
                        this._text_4
                    ], [], []);
                    return null;
                };
                return _View_ModalBodyComponent0;
            }(import1.AppView));
        }
    }
});
System.register("ng2-bs3-modal/components/modal-footer", ['@angular/core', "ng2-bs3-modal/components/modal"], function(exports_5, context_5) {
    "use strict";
    var __moduleName = context_5 && context_5.id;
    var core_3, modal_1;
    var ModalFooterComponent;
    return {
        setters:[
            function (core_3_1) {
                core_3 = core_3_1;
            },
            function (modal_1_1) {
                modal_1 = modal_1_1;
            }],
        execute: function() {
            ModalFooterComponent = (function () {
                function ModalFooterComponent(modal) {
                    this.modal = modal;
                    this.showDefaultButtons = false;
                    this.dismissButtonLabel = 'Dismiss';
                    this.closeButtonLabel = 'Close';
                }
                __decorate([
                    core_3.Input('show-default-buttons'), 
                    __metadata('design:type', Boolean)
                ], ModalFooterComponent.prototype, "showDefaultButtons", void 0);
                __decorate([
                    core_3.Input('dismiss-button-label'), 
                    __metadata('design:type', String)
                ], ModalFooterComponent.prototype, "dismissButtonLabel", void 0);
                __decorate([
                    core_3.Input('close-button-label'), 
                    __metadata('design:type', String)
                ], ModalFooterComponent.prototype, "closeButtonLabel", void 0);
                ModalFooterComponent = __decorate([
                    core_3.Component({
                        selector: 'modal-footer',
                        template: "\n        <div class=\"modal-footer\">\n            <ng-content></ng-content>\n            <button *ngIf=\"showDefaultButtons\" type=\"button\" class=\"btn btn-default\" data-dismiss=\"modal\" (click)=\"modal.dismiss()\">{{dismissButtonLabel}}</button>\n            <button *ngIf=\"showDefaultButtons\" type=\"button\" class=\"btn btn-primary\" (click)=\"modal.close()\">{{closeButtonLabel}}</button>\n        </div>\n    "
                    }), 
                    __metadata('design:paramtypes', [modal_1.ModalComponent])
                ], ModalFooterComponent);
                return ModalFooterComponent;
            }());
            exports_5("ModalFooterComponent", ModalFooterComponent);
        }
    }
});
/**
 * This file is generated by the Angular 2 template compiler.
 * Do not edit.
 */
/* tslint:disable */
System.register("ng2-bs3-modal/components/modal-footer.ngfactory", ['@angular/core/src/linker/view', '@angular/core/src/linker/element', "ng2-bs3-modal/components/modal-footer", '@angular/core/src/linker/view_utils', '@angular/core/src/linker/view_type', '@angular/core/src/change_detection/change_detection', "ng2-bs3-modal/components/modal", '@angular/core/src/metadata/view', '@angular/core/src/linker/component_factory', '@angular/common/src/directives/ng_if', '@angular/core/src/linker/template_ref'], function(exports_6, context_6) {
    "use strict";
    var __moduleName = context_6 && context_6.id;
    var import1, import2, import3, import4, import6, import7, import8, import9, import10, import11, import12;
    var renderType_ModalFooterComponent_Host, _View_ModalFooterComponent_Host0, ModalFooterComponentNgFactory, styles_ModalFooterComponent, renderType_ModalFooterComponent, _View_ModalFooterComponent0, _View_ModalFooterComponent1, _View_ModalFooterComponent2;
    function viewFactory_ModalFooterComponent_Host0(viewUtils, parentInjector, declarationEl) {
        if ((renderType_ModalFooterComponent_Host === null)) {
            (renderType_ModalFooterComponent_Host = viewUtils.createRenderComponentType('', 0, import9.ViewEncapsulation.None, [], {}));
        }
        return new _View_ModalFooterComponent_Host0(viewUtils, parentInjector, declarationEl);
    }
    function viewFactory_ModalFooterComponent0(viewUtils, parentInjector, declarationEl) {
        if ((renderType_ModalFooterComponent === null)) {
            (renderType_ModalFooterComponent = viewUtils.createRenderComponentType('/Users/doug/Source/dougludlow/ng2-bs3-modal/src/ng2-bs3-modal/components/modal-footer.ts class ModalFooterComponent - inline template', 1, import9.ViewEncapsulation.None, styles_ModalFooterComponent, {}));
        }
        return new _View_ModalFooterComponent0(viewUtils, parentInjector, declarationEl);
    }
    exports_6("viewFactory_ModalFooterComponent0", viewFactory_ModalFooterComponent0);
    function viewFactory_ModalFooterComponent1(viewUtils, parentInjector, declarationEl) {
        return new _View_ModalFooterComponent1(viewUtils, parentInjector, declarationEl);
    }
    function viewFactory_ModalFooterComponent2(viewUtils, parentInjector, declarationEl) {
        return new _View_ModalFooterComponent2(viewUtils, parentInjector, declarationEl);
    }
    return {
        setters:[
            function (import1_2) {
                import1 = import1_2;
            },
            function (import2_2) {
                import2 = import2_2;
            },
            function (import3_2) {
                import3 = import3_2;
            },
            function (import4_2) {
                import4 = import4_2;
            },
            function (import6_2) {
                import6 = import6_2;
            },
            function (import7_2) {
                import7 = import7_2;
            },
            function (import8_2) {
                import8 = import8_2;
            },
            function (import9_2) {
                import9 = import9_2;
            },
            function (import10_1) {
                import10 = import10_1;
            },
            function (import11_1) {
                import11 = import11_1;
            },
            function (import12_1) {
                import12 = import12_1;
            }],
        execute: function() {
            renderType_ModalFooterComponent_Host = null;
            _View_ModalFooterComponent_Host0 = (function (_super) {
                __extends(_View_ModalFooterComponent_Host0, _super);
                function _View_ModalFooterComponent_Host0(viewUtils, parentInjector, declarationEl) {
                    _super.call(this, _View_ModalFooterComponent_Host0, renderType_ModalFooterComponent_Host, import6.ViewType.HOST, viewUtils, parentInjector, declarationEl, import7.ChangeDetectorStatus.CheckAlways);
                }
                _View_ModalFooterComponent_Host0.prototype.createInternal = function (rootSelector) {
                    this._el_0 = this.selectOrCreateHostElement('modal-footer', rootSelector, null);
                    this._appEl_0 = new import2.AppElement(0, null, this, this._el_0);
                    var compView_0 = viewFactory_ModalFooterComponent0(this.viewUtils, this.injector(0), this._appEl_0);
                    this._ModalFooterComponent_0_4 = new import3.ModalFooterComponent(this.parentInjector.get(import8.ModalComponent));
                    this._appEl_0.initComponent(this._ModalFooterComponent_0_4, [], compView_0);
                    compView_0.create(this._ModalFooterComponent_0_4, this.projectableNodes, null);
                    this.init([].concat([this._el_0]), [this._el_0], [], []);
                    return this._appEl_0;
                };
                _View_ModalFooterComponent_Host0.prototype.injectorGetInternal = function (token, requestNodeIndex, notFoundResult) {
                    if (((token === import3.ModalFooterComponent) && (0 === requestNodeIndex))) {
                        return this._ModalFooterComponent_0_4;
                    }
                    return notFoundResult;
                };
                return _View_ModalFooterComponent_Host0;
            }(import1.AppView));
            exports_6("ModalFooterComponentNgFactory", ModalFooterComponentNgFactory = new import10.ComponentFactory('modal-footer', viewFactory_ModalFooterComponent_Host0, import3.ModalFooterComponent));
            styles_ModalFooterComponent = [];
            renderType_ModalFooterComponent = null;
            _View_ModalFooterComponent0 = (function (_super) {
                __extends(_View_ModalFooterComponent0, _super);
                function _View_ModalFooterComponent0(viewUtils, parentInjector, declarationEl) {
                    _super.call(this, _View_ModalFooterComponent0, renderType_ModalFooterComponent, import6.ViewType.COMPONENT, viewUtils, parentInjector, declarationEl, import7.ChangeDetectorStatus.CheckAlways);
                }
                _View_ModalFooterComponent0.prototype.createInternal = function (rootSelector) {
                    var parentRenderNode = this.renderer.createViewRoot(this.declarationAppElement.nativeElement);
                    this._text_0 = this.renderer.createText(parentRenderNode, '\n        ', null);
                    this._el_1 = this.renderer.createElement(parentRenderNode, 'div', null);
                    this.renderer.setElementAttribute(this._el_1, 'class', 'modal-footer');
                    this._text_2 = this.renderer.createText(this._el_1, '\n            ', null);
                    this.renderer.projectNodes(this._el_1, import4.flattenNestedViewRenderNodes(this.projectableNodes[0]));
                    this._text_3 = this.renderer.createText(this._el_1, '\n            ', null);
                    this._anchor_4 = this.renderer.createTemplateAnchor(this._el_1, null);
                    this._appEl_4 = new import2.AppElement(4, 1, this, this._anchor_4);
                    this._TemplateRef_4_5 = new import12.TemplateRef_(this._appEl_4, viewFactory_ModalFooterComponent1);
                    this._NgIf_4_6 = new import11.NgIf(this._appEl_4.vcRef, this._TemplateRef_4_5);
                    this._text_5 = this.renderer.createText(this._el_1, '\n            ', null);
                    this._anchor_6 = this.renderer.createTemplateAnchor(this._el_1, null);
                    this._appEl_6 = new import2.AppElement(6, 1, this, this._anchor_6);
                    this._TemplateRef_6_5 = new import12.TemplateRef_(this._appEl_6, viewFactory_ModalFooterComponent2);
                    this._NgIf_6_6 = new import11.NgIf(this._appEl_6.vcRef, this._TemplateRef_6_5);
                    this._text_7 = this.renderer.createText(this._el_1, '\n        ', null);
                    this._text_8 = this.renderer.createText(parentRenderNode, '\n    ', null);
                    this._expr_0 = import7.UNINITIALIZED;
                    this._expr_1 = import7.UNINITIALIZED;
                    this.init([], [
                        this._text_0,
                        this._el_1,
                        this._text_2,
                        this._text_3,
                        this._anchor_4,
                        this._text_5,
                        this._anchor_6,
                        this._text_7,
                        this._text_8
                    ], [], []);
                    return null;
                };
                _View_ModalFooterComponent0.prototype.injectorGetInternal = function (token, requestNodeIndex, notFoundResult) {
                    if (((token === import12.TemplateRef) && (4 === requestNodeIndex))) {
                        return this._TemplateRef_4_5;
                    }
                    if (((token === import11.NgIf) && (4 === requestNodeIndex))) {
                        return this._NgIf_4_6;
                    }
                    if (((token === import12.TemplateRef) && (6 === requestNodeIndex))) {
                        return this._TemplateRef_6_5;
                    }
                    if (((token === import11.NgIf) && (6 === requestNodeIndex))) {
                        return this._NgIf_6_6;
                    }
                    return notFoundResult;
                };
                _View_ModalFooterComponent0.prototype.detectChangesInternal = function (throwOnChange) {
                    var currVal_0 = this.context.showDefaultButtons;
                    if (import4.checkBinding(throwOnChange, this._expr_0, currVal_0)) {
                        this._NgIf_4_6.ngIf = currVal_0;
                        this._expr_0 = currVal_0;
                    }
                    var currVal_1 = this.context.showDefaultButtons;
                    if (import4.checkBinding(throwOnChange, this._expr_1, currVal_1)) {
                        this._NgIf_6_6.ngIf = currVal_1;
                        this._expr_1 = currVal_1;
                    }
                    this.detectContentChildrenChanges(throwOnChange);
                    this.detectViewChildrenChanges(throwOnChange);
                };
                return _View_ModalFooterComponent0;
            }(import1.AppView));
            _View_ModalFooterComponent1 = (function (_super) {
                __extends(_View_ModalFooterComponent1, _super);
                function _View_ModalFooterComponent1(viewUtils, parentInjector, declarationEl) {
                    _super.call(this, _View_ModalFooterComponent1, renderType_ModalFooterComponent, import6.ViewType.EMBEDDED, viewUtils, parentInjector, declarationEl, import7.ChangeDetectorStatus.CheckAlways);
                }
                _View_ModalFooterComponent1.prototype.createInternal = function (rootSelector) {
                    this._el_0 = this.renderer.createElement(null, 'button', null);
                    this.renderer.setElementAttribute(this._el_0, 'class', 'btn btn-default');
                    this.renderer.setElementAttribute(this._el_0, 'data-dismiss', 'modal');
                    this.renderer.setElementAttribute(this._el_0, 'type', 'button');
                    this._text_1 = this.renderer.createText(this._el_0, '', null);
                    var disposable_0 = this.renderer.listen(this._el_0, 'click', this.eventHandler(this._handle_click_0_0.bind(this)));
                    this._expr_1 = import7.UNINITIALIZED;
                    this.init([].concat([this._el_0]), [
                        this._el_0,
                        this._text_1
                    ], [disposable_0], []);
                    return null;
                };
                _View_ModalFooterComponent1.prototype.detectChangesInternal = function (throwOnChange) {
                    this.detectContentChildrenChanges(throwOnChange);
                    var currVal_1 = import4.interpolate(1, '', this.parent.context.dismissButtonLabel, '');
                    if (import4.checkBinding(throwOnChange, this._expr_1, currVal_1)) {
                        this.renderer.setText(this._text_1, currVal_1);
                        this._expr_1 = currVal_1;
                    }
                    this.detectViewChildrenChanges(throwOnChange);
                };
                _View_ModalFooterComponent1.prototype._handle_click_0_0 = function ($event) {
                    this.markPathToRootAsCheckOnce();
                    var pd_0 = (this.parent.context.modal.dismiss() !== false);
                    return (true && pd_0);
                };
                return _View_ModalFooterComponent1;
            }(import1.AppView));
            _View_ModalFooterComponent2 = (function (_super) {
                __extends(_View_ModalFooterComponent2, _super);
                function _View_ModalFooterComponent2(viewUtils, parentInjector, declarationEl) {
                    _super.call(this, _View_ModalFooterComponent2, renderType_ModalFooterComponent, import6.ViewType.EMBEDDED, viewUtils, parentInjector, declarationEl, import7.ChangeDetectorStatus.CheckAlways);
                }
                _View_ModalFooterComponent2.prototype.createInternal = function (rootSelector) {
                    this._el_0 = this.renderer.createElement(null, 'button', null);
                    this.renderer.setElementAttribute(this._el_0, 'class', 'btn btn-primary');
                    this.renderer.setElementAttribute(this._el_0, 'type', 'button');
                    this._text_1 = this.renderer.createText(this._el_0, '', null);
                    var disposable_0 = this.renderer.listen(this._el_0, 'click', this.eventHandler(this._handle_click_0_0.bind(this)));
                    this._expr_1 = import7.UNINITIALIZED;
                    this.init([].concat([this._el_0]), [
                        this._el_0,
                        this._text_1
                    ], [disposable_0], []);
                    return null;
                };
                _View_ModalFooterComponent2.prototype.detectChangesInternal = function (throwOnChange) {
                    this.detectContentChildrenChanges(throwOnChange);
                    var currVal_1 = import4.interpolate(1, '', this.parent.context.closeButtonLabel, '');
                    if (import4.checkBinding(throwOnChange, this._expr_1, currVal_1)) {
                        this.renderer.setText(this._text_1, currVal_1);
                        this._expr_1 = currVal_1;
                    }
                    this.detectViewChildrenChanges(throwOnChange);
                };
                _View_ModalFooterComponent2.prototype._handle_click_0_0 = function ($event) {
                    this.markPathToRootAsCheckOnce();
                    var pd_0 = (this.parent.context.modal.close() !== false);
                    return (true && pd_0);
                };
                return _View_ModalFooterComponent2;
            }(import1.AppView));
        }
    }
});
System.register("ng2-bs3-modal/components/modal-header", ['@angular/core', "ng2-bs3-modal/components/modal"], function(exports_7, context_7) {
    "use strict";
    var __moduleName = context_7 && context_7.id;
    var core_4, modal_2;
    var ModalHeaderComponent;
    return {
        setters:[
            function (core_4_1) {
                core_4 = core_4_1;
            },
            function (modal_2_1) {
                modal_2 = modal_2_1;
            }],
        execute: function() {
            ModalHeaderComponent = (function () {
                function ModalHeaderComponent(modal) {
                    this.modal = modal;
                    this.showClose = false;
                }
                __decorate([
                    core_4.Input('show-close'), 
                    __metadata('design:type', Boolean)
                ], ModalHeaderComponent.prototype, "showClose", void 0);
                ModalHeaderComponent = __decorate([
                    core_4.Component({
                        selector: 'modal-header',
                        template: "\n        <div class=\"modal-header\">\n            <button *ngIf=\"showClose\" type=\"button\" class=\"close\" data-dismiss=\"modal\" aria-label=\"Close\" (click)=\"modal.dismiss()\">\n                <span aria-hidden=\"true\">&times;</span>\n            </button>\n            <ng-content></ng-content>\n        </div>\n    "
                    }), 
                    __metadata('design:paramtypes', [modal_2.ModalComponent])
                ], ModalHeaderComponent);
                return ModalHeaderComponent;
            }());
            exports_7("ModalHeaderComponent", ModalHeaderComponent);
        }
    }
});
/**
 * This file is generated by the Angular 2 template compiler.
 * Do not edit.
 */
/* tslint:disable */
System.register("ng2-bs3-modal/components/modal-header.ngfactory", ['@angular/core/src/linker/view', '@angular/core/src/linker/element', "ng2-bs3-modal/components/modal-header", '@angular/core/src/linker/view_utils', '@angular/core/src/linker/view_type', '@angular/core/src/change_detection/change_detection', "ng2-bs3-modal/components/modal", '@angular/core/src/metadata/view', '@angular/core/src/linker/component_factory', '@angular/common/src/directives/ng_if', '@angular/core/src/linker/template_ref'], function(exports_8, context_8) {
    "use strict";
    var __moduleName = context_8 && context_8.id;
    var import1, import2, import3, import4, import6, import7, import8, import9, import10, import11, import12;
    var renderType_ModalHeaderComponent_Host, _View_ModalHeaderComponent_Host0, ModalHeaderComponentNgFactory, styles_ModalHeaderComponent, renderType_ModalHeaderComponent, _View_ModalHeaderComponent0, _View_ModalHeaderComponent1;
    function viewFactory_ModalHeaderComponent_Host0(viewUtils, parentInjector, declarationEl) {
        if ((renderType_ModalHeaderComponent_Host === null)) {
            (renderType_ModalHeaderComponent_Host = viewUtils.createRenderComponentType('', 0, import9.ViewEncapsulation.None, [], {}));
        }
        return new _View_ModalHeaderComponent_Host0(viewUtils, parentInjector, declarationEl);
    }
    function viewFactory_ModalHeaderComponent0(viewUtils, parentInjector, declarationEl) {
        if ((renderType_ModalHeaderComponent === null)) {
            (renderType_ModalHeaderComponent = viewUtils.createRenderComponentType('/Users/doug/Source/dougludlow/ng2-bs3-modal/src/ng2-bs3-modal/components/modal-header.ts class ModalHeaderComponent - inline template', 1, import9.ViewEncapsulation.None, styles_ModalHeaderComponent, {}));
        }
        return new _View_ModalHeaderComponent0(viewUtils, parentInjector, declarationEl);
    }
    exports_8("viewFactory_ModalHeaderComponent0", viewFactory_ModalHeaderComponent0);
    function viewFactory_ModalHeaderComponent1(viewUtils, parentInjector, declarationEl) {
        return new _View_ModalHeaderComponent1(viewUtils, parentInjector, declarationEl);
    }
    return {
        setters:[
            function (import1_3) {
                import1 = import1_3;
            },
            function (import2_3) {
                import2 = import2_3;
            },
            function (import3_3) {
                import3 = import3_3;
            },
            function (import4_3) {
                import4 = import4_3;
            },
            function (import6_3) {
                import6 = import6_3;
            },
            function (import7_3) {
                import7 = import7_3;
            },
            function (import8_3) {
                import8 = import8_3;
            },
            function (import9_3) {
                import9 = import9_3;
            },
            function (import10_2) {
                import10 = import10_2;
            },
            function (import11_2) {
                import11 = import11_2;
            },
            function (import12_2) {
                import12 = import12_2;
            }],
        execute: function() {
            renderType_ModalHeaderComponent_Host = null;
            _View_ModalHeaderComponent_Host0 = (function (_super) {
                __extends(_View_ModalHeaderComponent_Host0, _super);
                function _View_ModalHeaderComponent_Host0(viewUtils, parentInjector, declarationEl) {
                    _super.call(this, _View_ModalHeaderComponent_Host0, renderType_ModalHeaderComponent_Host, import6.ViewType.HOST, viewUtils, parentInjector, declarationEl, import7.ChangeDetectorStatus.CheckAlways);
                }
                _View_ModalHeaderComponent_Host0.prototype.createInternal = function (rootSelector) {
                    this._el_0 = this.selectOrCreateHostElement('modal-header', rootSelector, null);
                    this._appEl_0 = new import2.AppElement(0, null, this, this._el_0);
                    var compView_0 = viewFactory_ModalHeaderComponent0(this.viewUtils, this.injector(0), this._appEl_0);
                    this._ModalHeaderComponent_0_4 = new import3.ModalHeaderComponent(this.parentInjector.get(import8.ModalComponent));
                    this._appEl_0.initComponent(this._ModalHeaderComponent_0_4, [], compView_0);
                    compView_0.create(this._ModalHeaderComponent_0_4, this.projectableNodes, null);
                    this.init([].concat([this._el_0]), [this._el_0], [], []);
                    return this._appEl_0;
                };
                _View_ModalHeaderComponent_Host0.prototype.injectorGetInternal = function (token, requestNodeIndex, notFoundResult) {
                    if (((token === import3.ModalHeaderComponent) && (0 === requestNodeIndex))) {
                        return this._ModalHeaderComponent_0_4;
                    }
                    return notFoundResult;
                };
                return _View_ModalHeaderComponent_Host0;
            }(import1.AppView));
            exports_8("ModalHeaderComponentNgFactory", ModalHeaderComponentNgFactory = new import10.ComponentFactory('modal-header', viewFactory_ModalHeaderComponent_Host0, import3.ModalHeaderComponent));
            styles_ModalHeaderComponent = [];
            renderType_ModalHeaderComponent = null;
            _View_ModalHeaderComponent0 = (function (_super) {
                __extends(_View_ModalHeaderComponent0, _super);
                function _View_ModalHeaderComponent0(viewUtils, parentInjector, declarationEl) {
                    _super.call(this, _View_ModalHeaderComponent0, renderType_ModalHeaderComponent, import6.ViewType.COMPONENT, viewUtils, parentInjector, declarationEl, import7.ChangeDetectorStatus.CheckAlways);
                }
                _View_ModalHeaderComponent0.prototype.createInternal = function (rootSelector) {
                    var parentRenderNode = this.renderer.createViewRoot(this.declarationAppElement.nativeElement);
                    this._text_0 = this.renderer.createText(parentRenderNode, '\n        ', null);
                    this._el_1 = this.renderer.createElement(parentRenderNode, 'div', null);
                    this.renderer.setElementAttribute(this._el_1, 'class', 'modal-header');
                    this._text_2 = this.renderer.createText(this._el_1, '\n            ', null);
                    this._anchor_3 = this.renderer.createTemplateAnchor(this._el_1, null);
                    this._appEl_3 = new import2.AppElement(3, 1, this, this._anchor_3);
                    this._TemplateRef_3_5 = new import12.TemplateRef_(this._appEl_3, viewFactory_ModalHeaderComponent1);
                    this._NgIf_3_6 = new import11.NgIf(this._appEl_3.vcRef, this._TemplateRef_3_5);
                    this._text_4 = this.renderer.createText(this._el_1, '\n            ', null);
                    this.renderer.projectNodes(this._el_1, import4.flattenNestedViewRenderNodes(this.projectableNodes[0]));
                    this._text_5 = this.renderer.createText(this._el_1, '\n        ', null);
                    this._text_6 = this.renderer.createText(parentRenderNode, '\n    ', null);
                    this._expr_0 = import7.UNINITIALIZED;
                    this.init([], [
                        this._text_0,
                        this._el_1,
                        this._text_2,
                        this._anchor_3,
                        this._text_4,
                        this._text_5,
                        this._text_6
                    ], [], []);
                    return null;
                };
                _View_ModalHeaderComponent0.prototype.injectorGetInternal = function (token, requestNodeIndex, notFoundResult) {
                    if (((token === import12.TemplateRef) && (3 === requestNodeIndex))) {
                        return this._TemplateRef_3_5;
                    }
                    if (((token === import11.NgIf) && (3 === requestNodeIndex))) {
                        return this._NgIf_3_6;
                    }
                    return notFoundResult;
                };
                _View_ModalHeaderComponent0.prototype.detectChangesInternal = function (throwOnChange) {
                    var currVal_0 = this.context.showClose;
                    if (import4.checkBinding(throwOnChange, this._expr_0, currVal_0)) {
                        this._NgIf_3_6.ngIf = currVal_0;
                        this._expr_0 = currVal_0;
                    }
                    this.detectContentChildrenChanges(throwOnChange);
                    this.detectViewChildrenChanges(throwOnChange);
                };
                return _View_ModalHeaderComponent0;
            }(import1.AppView));
            _View_ModalHeaderComponent1 = (function (_super) {
                __extends(_View_ModalHeaderComponent1, _super);
                function _View_ModalHeaderComponent1(viewUtils, parentInjector, declarationEl) {
                    _super.call(this, _View_ModalHeaderComponent1, renderType_ModalHeaderComponent, import6.ViewType.EMBEDDED, viewUtils, parentInjector, declarationEl, import7.ChangeDetectorStatus.CheckAlways);
                }
                _View_ModalHeaderComponent1.prototype.createInternal = function (rootSelector) {
                    this._el_0 = this.renderer.createElement(null, 'button', null);
                    this.renderer.setElementAttribute(this._el_0, 'aria-label', 'Close');
                    this.renderer.setElementAttribute(this._el_0, 'class', 'close');
                    this.renderer.setElementAttribute(this._el_0, 'data-dismiss', 'modal');
                    this.renderer.setElementAttribute(this._el_0, 'type', 'button');
                    this._text_1 = this.renderer.createText(this._el_0, '\n                ', null);
                    this._el_2 = this.renderer.createElement(this._el_0, 'span', null);
                    this.renderer.setElementAttribute(this._el_2, 'aria-hidden', 'true');
                    this._text_3 = this.renderer.createText(this._el_2, '', null);
                    this._text_4 = this.renderer.createText(this._el_0, '\n            ', null);
                    var disposable_0 = this.renderer.listen(this._el_0, 'click', this.eventHandler(this._handle_click_0_0.bind(this)));
                    this.init([].concat([this._el_0]), [
                        this._el_0,
                        this._text_1,
                        this._el_2,
                        this._text_3,
                        this._text_4
                    ], [disposable_0], []);
                    return null;
                };
                _View_ModalHeaderComponent1.prototype._handle_click_0_0 = function ($event) {
                    this.markPathToRootAsCheckOnce();
                    var pd_0 = (this.parent.context.modal.dismiss() !== false);
                    return (true && pd_0);
                };
                return _View_ModalHeaderComponent1;
            }(import1.AppView));
        }
    }
});
/**
 * This file is generated by the Angular 2 template compiler.
 * Do not edit.
 */
/* tslint:disable */
System.register("ng2-bs3-modal/components/modal.ngfactory", ['@angular/core/src/linker/view', '@angular/core/src/linker/element', "ng2-bs3-modal/components/modal", '@angular/core/src/linker/view_utils', '@angular/core/src/linker/view_type', '@angular/core/src/change_detection/change_detection', '@angular/core/src/linker/element_ref', '@angular/core/src/metadata/view', '@angular/core/src/linker/component_factory', '@angular/common/src/directives/ng_class', '@angular/core/src/change_detection/differs/iterable_differs', '@angular/core/src/change_detection/differs/keyvalue_differs'], function(exports_9, context_9) {
    "use strict";
    var __moduleName = context_9 && context_9.id;
    var import1, import2, import3, import4, import6, import7, import8, import9, import10, import11, import12, import13;
    var renderType_ModalComponent_Host, _View_ModalComponent_Host0, ModalComponentNgFactory, styles_ModalComponent, renderType_ModalComponent, _View_ModalComponent0;
    function viewFactory_ModalComponent_Host0(viewUtils, parentInjector, declarationEl) {
        if ((renderType_ModalComponent_Host === null)) {
            (renderType_ModalComponent_Host = viewUtils.createRenderComponentType('', 0, import9.ViewEncapsulation.None, [], {}));
        }
        return new _View_ModalComponent_Host0(viewUtils, parentInjector, declarationEl);
    }
    function viewFactory_ModalComponent0(viewUtils, parentInjector, declarationEl) {
        if ((renderType_ModalComponent === null)) {
            (renderType_ModalComponent = viewUtils.createRenderComponentType('/Users/doug/Source/dougludlow/ng2-bs3-modal/src/ng2-bs3-modal/components/modal.ts class ModalComponent - inline template', 1, import9.ViewEncapsulation.None, styles_ModalComponent, {}));
        }
        return new _View_ModalComponent0(viewUtils, parentInjector, declarationEl);
    }
    exports_9("viewFactory_ModalComponent0", viewFactory_ModalComponent0);
    return {
        setters:[
            function (import1_4) {
                import1 = import1_4;
            },
            function (import2_4) {
                import2 = import2_4;
            },
            function (import3_4) {
                import3 = import3_4;
            },
            function (import4_4) {
                import4 = import4_4;
            },
            function (import6_4) {
                import6 = import6_4;
            },
            function (import7_4) {
                import7 = import7_4;
            },
            function (import8_4) {
                import8 = import8_4;
            },
            function (import9_4) {
                import9 = import9_4;
            },
            function (import10_3) {
                import10 = import10_3;
            },
            function (import11_3) {
                import11 = import11_3;
            },
            function (import12_3) {
                import12 = import12_3;
            },
            function (import13_1) {
                import13 = import13_1;
            }],
        execute: function() {
            renderType_ModalComponent_Host = null;
            _View_ModalComponent_Host0 = (function (_super) {
                __extends(_View_ModalComponent_Host0, _super);
                function _View_ModalComponent_Host0(viewUtils, parentInjector, declarationEl) {
                    _super.call(this, _View_ModalComponent_Host0, renderType_ModalComponent_Host, import6.ViewType.HOST, viewUtils, parentInjector, declarationEl, import7.ChangeDetectorStatus.CheckAlways);
                }
                _View_ModalComponent_Host0.prototype.createInternal = function (rootSelector) {
                    this._el_0 = this.selectOrCreateHostElement('modal', rootSelector, null);
                    this.renderer.setElementAttribute(this._el_0, 'class', 'modal');
                    this.renderer.setElementAttribute(this._el_0, 'role', 'dialog');
                    this.renderer.setElementAttribute(this._el_0, 'tabindex', '-1');
                    this._appEl_0 = new import2.AppElement(0, null, this, this._el_0);
                    var compView_0 = viewFactory_ModalComponent0(this.viewUtils, this.injector(0), this._appEl_0);
                    this._ModalComponent_0_4 = new import3.ModalComponent(new import8.ElementRef(this._el_0));
                    this._appEl_0.initComponent(this._ModalComponent_0_4, [], compView_0);
                    compView_0.create(this._ModalComponent_0_4, this.projectableNodes, null);
                    this._expr_0 = import7.UNINITIALIZED;
                    this._expr_1 = import7.UNINITIALIZED;
                    this._expr_2 = import7.UNINITIALIZED;
                    this.init([].concat([this._el_0]), [this._el_0], [], []);
                    return this._appEl_0;
                };
                _View_ModalComponent_Host0.prototype.injectorGetInternal = function (token, requestNodeIndex, notFoundResult) {
                    if (((token === import3.ModalComponent) && (0 === requestNodeIndex))) {
                        return this._ModalComponent_0_4;
                    }
                    return notFoundResult;
                };
                _View_ModalComponent_Host0.prototype.detectChangesInternal = function (throwOnChange) {
                    this.detectContentChildrenChanges(throwOnChange);
                    var currVal_0 = this._ModalComponent_0_4.fadeClass;
                    if (import4.checkBinding(throwOnChange, this._expr_0, currVal_0)) {
                        this.renderer.setElementClass(this._el_0, 'fade', currVal_0);
                        this._expr_0 = currVal_0;
                    }
                    var currVal_1 = this._ModalComponent_0_4.dataKeyboardAttr;
                    if (import4.checkBinding(throwOnChange, this._expr_1, currVal_1)) {
                        this.renderer.setElementAttribute(this._el_0, 'data-keyboard', ((currVal_1 == null) ? null : currVal_1.toString()));
                        this._expr_1 = currVal_1;
                    }
                    var currVal_2 = this._ModalComponent_0_4.dataBackdropAttr;
                    if (import4.checkBinding(throwOnChange, this._expr_2, currVal_2)) {
                        this.renderer.setElementAttribute(this._el_0, 'data-backdrop', ((currVal_2 == null) ? null : currVal_2.toString()));
                        this._expr_2 = currVal_2;
                    }
                    this.detectViewChildrenChanges(throwOnChange);
                };
                _View_ModalComponent_Host0.prototype.destroyInternal = function () {
                    this._ModalComponent_0_4.ngOnDestroy();
                };
                return _View_ModalComponent_Host0;
            }(import1.AppView));
            exports_9("ModalComponentNgFactory", ModalComponentNgFactory = new import10.ComponentFactory('modal', viewFactory_ModalComponent_Host0, import3.ModalComponent));
            styles_ModalComponent = [];
            renderType_ModalComponent = null;
            _View_ModalComponent0 = (function (_super) {
                __extends(_View_ModalComponent0, _super);
                function _View_ModalComponent0(viewUtils, parentInjector, declarationEl) {
                    _super.call(this, _View_ModalComponent0, renderType_ModalComponent, import6.ViewType.COMPONENT, viewUtils, parentInjector, declarationEl, import7.ChangeDetectorStatus.CheckAlways);
                }
                _View_ModalComponent0.prototype.createInternal = function (rootSelector) {
                    var parentRenderNode = this.renderer.createViewRoot(this.declarationAppElement.nativeElement);
                    this._text_0 = this.renderer.createText(parentRenderNode, '\n        ', null);
                    this._el_1 = this.renderer.createElement(parentRenderNode, 'div', null);
                    this.renderer.setElementAttribute(this._el_1, 'class', 'modal-dialog');
                    this._NgClass_1_3 = new import11.NgClass(this.parentInjector.get(import12.IterableDiffers), this.parentInjector.get(import13.KeyValueDiffers), new import8.ElementRef(this._el_1), this.renderer);
                    this._text_2 = this.renderer.createText(this._el_1, '\n            ', null);
                    this._el_3 = this.renderer.createElement(this._el_1, 'div', null);
                    this.renderer.setElementAttribute(this._el_3, 'class', 'modal-content');
                    this._text_4 = this.renderer.createText(this._el_3, '\n                ', null);
                    this.renderer.projectNodes(this._el_3, import4.flattenNestedViewRenderNodes(this.projectableNodes[0]));
                    this._text_5 = this.renderer.createText(this._el_3, '\n            ', null);
                    this._text_6 = this.renderer.createText(this._el_1, '\n        ', null);
                    this._text_7 = this.renderer.createText(parentRenderNode, '\n    ', null);
                    this._expr_0 = import7.UNINITIALIZED;
                    this._expr_1 = import7.UNINITIALIZED;
                    this.init([], [
                        this._text_0,
                        this._el_1,
                        this._text_2,
                        this._el_3,
                        this._text_4,
                        this._text_5,
                        this._text_6,
                        this._text_7
                    ], [], []);
                    return null;
                };
                _View_ModalComponent0.prototype.injectorGetInternal = function (token, requestNodeIndex, notFoundResult) {
                    if (((token === import11.NgClass) && ((1 <= requestNodeIndex) && (requestNodeIndex <= 6)))) {
                        return this._NgClass_1_3;
                    }
                    return notFoundResult;
                };
                _View_ModalComponent0.prototype.detectChangesInternal = function (throwOnChange) {
                    var currVal_0 = 'modal-dialog';
                    if (import4.checkBinding(throwOnChange, this._expr_0, currVal_0)) {
                        this._NgClass_1_3.klass = currVal_0;
                        this._expr_0 = currVal_0;
                    }
                    var currVal_1 = this.context.getCssClasses();
                    if (import4.checkBinding(throwOnChange, this._expr_1, currVal_1)) {
                        this._NgClass_1_3.ngClass = currVal_1;
                        this._expr_1 = currVal_1;
                    }
                    if (!throwOnChange) {
                        this._NgClass_1_3.ngDoCheck();
                    }
                    this.detectContentChildrenChanges(throwOnChange);
                    this.detectViewChildrenChanges(throwOnChange);
                };
                return _View_ModalComponent0;
            }(import1.AppView));
        }
    }
});
System.register("ng2-bs3-modal/directives/autofocus", ['@angular/core', "ng2-bs3-modal/components/modal"], function(exports_10, context_10) {
    "use strict";
    var __moduleName = context_10 && context_10.id;
    var core_5, modal_3;
    var AutofocusDirective;
    return {
        setters:[
            function (core_5_1) {
                core_5 = core_5_1;
            },
            function (modal_3_1) {
                modal_3 = modal_3_1;
            }],
        execute: function() {
            AutofocusDirective = (function () {
                function AutofocusDirective(el, modal) {
                    var _this = this;
                    this.el = el;
                    this.modal = modal;
                    if (modal) {
                        this.modal.onOpen.subscribe(function () {
                            _this.el.nativeElement.focus();
                        });
                    }
                }
                AutofocusDirective = __decorate([
                    core_5.Directive({
                        selector: '[autofocus]'
                    }),
                    __param(1, core_5.Optional()), 
                    __metadata('design:paramtypes', [core_5.ElementRef, modal_3.ModalComponent])
                ], AutofocusDirective);
                return AutofocusDirective;
            }());
            exports_10("AutofocusDirective", AutofocusDirective);
        }
    }
});
System.register("ng2-bs3-modal/ng2-bs3-modal", ['@angular/core', '@angular/common', "ng2-bs3-modal/components/modal", "ng2-bs3-modal/components/modal-header", "ng2-bs3-modal/components/modal-body", "ng2-bs3-modal/components/modal-footer", "ng2-bs3-modal/directives/autofocus", "ng2-bs3-modal/components/modal-instance"], function(exports_11, context_11) {
    "use strict";
    var __moduleName = context_11 && context_11.id;
    var core_6, common_1, modal_4, modal_header_1, modal_body_1, modal_footer_1, autofocus_1;
    var Ng2Bs3ModalModule;
    var exportedNames_1 = {
        'Ng2Bs3ModalModule': true
    };
    function exportStar_1(m) {
        var exports = {};
        for(var n in m) {
            if (n !== "default"&& !exportedNames_1.hasOwnProperty(n)) exports[n] = m[n];
        }
        exports_11(exports);
    }
    return {
        setters:[
            function (core_6_1) {
                core_6 = core_6_1;
            },
            function (common_1_1) {
                common_1 = common_1_1;
            },
            function (modal_4_1) {
                modal_4 = modal_4_1;
                exportStar_1(modal_4_1);
            },
            function (modal_header_1_1) {
                modal_header_1 = modal_header_1_1;
                exportStar_1(modal_header_1_1);
            },
            function (modal_body_1_1) {
                modal_body_1 = modal_body_1_1;
                exportStar_1(modal_body_1_1);
            },
            function (modal_footer_1_1) {
                modal_footer_1 = modal_footer_1_1;
                exportStar_1(modal_footer_1_1);
            },
            function (autofocus_1_1) {
                autofocus_1 = autofocus_1_1;
            },
            function (modal_instance_2_1) {
                exportStar_1(modal_instance_2_1);
            }],
        execute: function() {
            Ng2Bs3ModalModule = (function () {
                function Ng2Bs3ModalModule() {
                }
                Ng2Bs3ModalModule = __decorate([
                    core_6.NgModule({
                        imports: [
                            common_1.CommonModule
                        ],
                        declarations: [
                            modal_4.ModalComponent,
                            modal_header_1.ModalHeaderComponent,
                            modal_body_1.ModalBodyComponent,
                            modal_footer_1.ModalFooterComponent,
                            autofocus_1.AutofocusDirective
                        ],
                        exports: [
                            modal_4.ModalComponent,
                            modal_header_1.ModalHeaderComponent,
                            modal_body_1.ModalBodyComponent,
                            modal_footer_1.ModalFooterComponent,
                            autofocus_1.AutofocusDirective
                        ]
                    }), 
                    __metadata('design:paramtypes', [])
                ], Ng2Bs3ModalModule);
                return Ng2Bs3ModalModule;
            }());
            exports_11("Ng2Bs3ModalModule", Ng2Bs3ModalModule);
        }
    }
});
/**
 * This file is generated by the Angular 2 template compiler.
 * Do not edit.
 */
/* tslint:disable */
System.register("ng2-bs3-modal/ng2-bs3-modal.ngfactory", ['@angular/core/src/linker/ng_module_factory', "ng2-bs3-modal/ng2-bs3-modal", '@angular/common/src/common_module', '@angular/common/src/localization', '@angular/core/src/i18n/tokens'], function(exports_12, context_12) {
    "use strict";
    var __moduleName = context_12 && context_12.id;
    var import0, import1, import2, import3, import5;
    var Ng2Bs3ModalModuleInjector, Ng2Bs3ModalModuleNgFactory;
    return {
        setters:[
            function (import0_1) {
                import0 = import0_1;
            },
            function (import1_5) {
                import1 = import1_5;
            },
            function (import2_5) {
                import2 = import2_5;
            },
            function (import3_5) {
                import3 = import3_5;
            },
            function (import5_1) {
                import5 = import5_1;
            }],
        execute: function() {
            Ng2Bs3ModalModuleInjector = (function (_super) {
                __extends(Ng2Bs3ModalModuleInjector, _super);
                function Ng2Bs3ModalModuleInjector(parent) {
                    _super.call(this, parent, [], []);
                }
                Object.defineProperty(Ng2Bs3ModalModuleInjector.prototype, "_NgLocalization_2", {
                    get: function () {
                        if ((this.__NgLocalization_2 == null)) {
                            (this.__NgLocalization_2 = new import3.NgLocaleLocalization(this.parent.get(import5.LOCALE_ID)));
                        }
                        return this.__NgLocalization_2;
                    },
                    enumerable: true,
                    configurable: true
                });
                Ng2Bs3ModalModuleInjector.prototype.createInternal = function () {
                    this._CommonModule_0 = new import2.CommonModule();
                    this._Ng2Bs3ModalModule_1 = new import1.Ng2Bs3ModalModule();
                    return this._Ng2Bs3ModalModule_1;
                };
                Ng2Bs3ModalModuleInjector.prototype.getInternal = function (token, notFoundResult) {
                    if ((token === import2.CommonModule)) {
                        return this._CommonModule_0;
                    }
                    if ((token === import1.Ng2Bs3ModalModule)) {
                        return this._Ng2Bs3ModalModule_1;
                    }
                    if ((token === import3.NgLocalization)) {
                        return this._NgLocalization_2;
                    }
                    return notFoundResult;
                };
                Ng2Bs3ModalModuleInjector.prototype.destroyInternal = function () {
                };
                return Ng2Bs3ModalModuleInjector;
            }(import0.NgModuleInjector));
            exports_12("Ng2Bs3ModalModuleNgFactory", Ng2Bs3ModalModuleNgFactory = new import0.NgModuleFactory(Ng2Bs3ModalModuleInjector, import1.Ng2Bs3ModalModule));
        }
    }
});
