// @angular-devkit/core/src/json/schema/serializers/javascript.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
const exception_1 = require("../../../exception/exception");
const strings_1 = require("../../../utils/strings");
const interface_1 = require("./interface");
class InvalidRangeException extends exception_1.BaseException {
    constructor(name, value, comparator, expected) {
        super(`Property ${JSON.stringify(name)} expected a value `
            + `${comparator} ${JSON.stringify(expected)}, received ${JSON.stringify(value)}.`);
    }
}
exports.InvalidRangeException = InvalidRangeException;
class InvalidValueException extends exception_1.BaseException {
    constructor(name, value, expected) {
        super(`Property ${JSON.stringify(name)} expected a value of type ${expected}, `
            + `received ${value}.`);
    }
}
exports.InvalidValueException = InvalidValueException;
class InvalidSchemaException extends exception_1.BaseException {
    constructor(schema) {
        super(`Invalid schema: ${JSON.stringify(schema)}`);
    }
}
exports.InvalidSchemaException = InvalidSchemaException;
class InvalidPropertyNameException extends exception_1.BaseException {
    constructor(path) {
        super(`Property ${JSON.stringify(path)} does not exist in the schema, and no additional `
            + `properties are allowed.`);
        this.path = path;
    }
}
exports.InvalidPropertyNameException = InvalidPropertyNameException;
class RequiredValueMissingException extends exception_1.BaseException {
    constructor(path) {
        super(`Property ${JSON.stringify(path)} is required but missing.`);
        this.path = path;
    }
}
exports.RequiredValueMissingException = RequiredValueMissingException;
exports.exceptions = {
    InvalidRangeException,
    InvalidSchemaException,
    InvalidValueException,
    InvalidPropertyNameException,
    RequiredValueMissingException,
};
const symbols = {
    Schema: Symbol('schema'),
};
class JavascriptSerializer extends interface_1.JsonSchemaSerializer {
    constructor(_options) {
        super();
        this._options = _options;
        this._uniqueSet = new Set();
    }
    _unique(name) {
        let i = 1;
        let result = name;
        while (this._uniqueSet.has(result)) {
            result = name + i;
            i++;
        }
        this._uniqueSet.add(result);
        return result;
    }
    serialize(ref, registry) {
        const rootSchema = registry.getSchemaFromRef(ref);
        const { root, templates } = require('./templates/javascript');
        const source = root({
            exceptions: exports.exceptions,
            name: '',
            options: this._options || {},
            schema: rootSchema,
            strings: {
                classify: strings_1.classify,
                camelize: strings_1.camelize,
            },
            symbols,
            templates,
        });
        const fn = new Function('registry', 'exceptions', 'symbols', 'value', source);
        return (value) => fn(registry, exports.exceptions, symbols, value);
    }
}
exports.JavascriptSerializer = JavascriptSerializer;
