// @angular-devkit/core/src/json/schema/serializers/templates/javascript/prop-array.js
return module.exports.default = function(obj) {
  obj || (obj = {});
  let __t;
  let __p = "";
  const __escapes = {"&":"&amp;","<":"&lt;",">":"&gt;","\"":"&quot;","'":"&#39;","`":"&#96;"};
  const __escapesre = new RegExp('[&<>"\\\'`]', 'g');

  const __e = function(s) {     return s ? s.replace(__escapesre, function(key) { return __escapes[key]; }) : '';  };
  with (obj) {
__p = __p
    + "(function() {\n"
    + "";

  if ('default' in schema && !Array.isArray(schema.default)
    || 'minItems' in schema && (typeof schema.minItems != 'number' || schema.minItems < 0)
    || 'maxItems' in schema && (typeof schema.maxItems != 'number' || schema.maxItems < 0)
    || 'uniqueItems' in schema && typeof schema.uniqueItems != 'boolean') {
    throw new exceptions.InvalidSchemaException(schema);
  }

  const required = (schema.required || []);
  const extras = {
    exceptions: exceptions,
    options: options,
    path: path,
    strings: strings,
    symbols: symbols,
    templates: templates,
  };

__p = __p
    + "\n"
    + "\n"
    + "const itemHandler = function() { return (";
__p += ((__t = 
  templates.subschema(Object.assign({
    name: '???',
    required: false,
    schema: schema.additionalProperties,
  }, extras))
) == null ? "" : __t);
__p = __p
    + "); };\n"
    + "\n"
    + "\n"
    + "const items = [];\n"
    + "const arrayFunctions = {\n"
    + "  get length() { return items.length; },\n"
    + "  push() { items.push.apply(items, arguments); },\n"
    + "  pop() { return items.pop(); },\n"
    + "  shift() { return items.shift(); },\n"
    + "  unshift() { return items.unshift(); },\n"
    + "  slice(start, end) { return items.slice(start, end); },\n"
    + "};\n"
    + "\n"
    + "\n"
    + "let defined = false;\n"
    + "const proxy = new Proxy({}, {\n"
    + "  isExtensible() { return false; },\n"
    + "  has(target, prop) {\n"
    + "    return (prop in items);\n"
    + "  },\n"
    + "  get(target, prop) {\n"
    + "    if (prop === symbols.Schema) {\n"
    + "      return arrayHandler.schema;\n"
    + "    }\n"
    + "\n"
    + "    if (prop >= 0 && prop in value) {\n"
    + "      return value[prop].get();\n"
    + "    }\n"
    + "    if (prop in arrayFunctions) {\n"
    + "      return arrayFunctions[prop];\n"
    + "    }\n"
    + "    return undefined;\n"
    + "  },\n"
    + "  set(target, prop, v) {\n"
    + "    if (prop >= 0) {\n"
    + "      if (!(prop in items)) {\n"
    + "        items[prop] = itemHandler();\n"
    + "      }\n"
    + "      items[prop].set(v);\n"
    + "      return true;\n"
    + "    }\n"
    + "    return false;\n"
    + "  },\n"
    + "  deleteProperty(target, prop) {\n"
    + "    if (prop >= 0 && prop in value) {\n"
    + "      value[prop].remove();\n"
    + "      return true;\n"
    + "    }\n"
    + "    return false;\n"
    + "  },\n"
    + "  defineProperty(target, prop, descriptor) {\n"
    + "    return false;\n"
    + "  },\n"
    + "  getOwnPropertyDescriptor(target, prop) {\n"
    + "    if (prop >= 0 && prop in value) {\n"
    + "      return { configurable: true, enumerable: true };\n"
    + "    }\n"
    + "  },\n"
    + "  ownKeys(target) {\n"
    + "    return Object.keys(items);\n"
    + "  },\n"
    + "});\n"
    + "\n"
    + "const arrayHandler = {\n"
    + "  set(v) {\n"
    + "    if (v === undefined) {\n"
    + "      defined = false;\n"
    + "      return;\n"
    + "    }\n"
    + "\n"
    + "    defined = true;\n"
    + "    for (const key of Object.keys(v)) {\n"
    + "      proxy[key] = v[key];\n"
    + "    }\n"
    + "\n"
    + "    // Validate required fields.\n"
    + "    ";
 for (const key of required) { 
__p = __p
    + "\n"
    + "    if (!(";
__p += ((__t =  JSON.stringify(key) ) == null ? "" : __t);
__p = __p
    + " in v)) {\n"
    + "      throw new exceptions.RequiredValueMissingException(";
__p += ((__t =  JSON.stringify(path) ) == null ? "" : __t);
__p = __p
    + " + '/' + ";
__p += ((__t =  JSON.stringify(key) ) == null ? "" : __t);
__p = __p
    + ");\n"
    + "    }";
 } 
__p = __p
    + "\n"
    + "  },\n"
    + "  get() {\n"
    + "    if (defined) {\n"
    + "      return proxy;\n"
    + "    } else {\n"
    + "      return ";
__p += ((__t =  'default' in schema ? JSON.stringify(schema.default) : 'undefined' ) == null ? "" : __t);
__p = __p
    + ";\n"
    + "    }\n"
    + "  },\n"
    + "  isDefined() { return defined; },\n"
    + "  remove() { this.set(undefined); },\n"
    + "  schema: ";
__p += ((__t =  JSON.stringify(schema) ) == null ? "" : __t);
__p = __p
    + ",\n"
    + "};\n"
    + "\n"
    + "return arrayHandler;\n"
    + "\n"
    + "})()\n"
    + "";
  };

  return __p;
}
