// @angular-devkit/core/src/json/schema/serializers/templates/javascript/prop-object.js
return module.exports.default = function(obj) {
  obj || (obj = {});
  let __t;
  let __p = "";
  const __escapes = {"&":"&amp;","<":"&lt;",">":"&gt;","\"":"&quot;","'":"&#39;","`":"&#96;"};
  const __escapesre = new RegExp('[&<>"\\\'`]', 'g');

  const __e = function(s) {     return s ? s.replace(__escapesre, function(key) { return __escapes[key]; }) : '';  };
  with (obj) {
__p = __p
    + "(function() {\n"
    + "";

  const required = (schema.required || []);
  const extras = {
    exceptions: exceptions,
    options: options,
    path: path,
    strings: strings,
    symbols: symbols,
    templates: templates,
  };

__p = __p
    + "\n"
    + "\n"
    + "const additionalProperties = {};\n"
    + "const additionalPropertyHandler = ";

if (!('additionalProperties' in schema)) { 
__p = __p
    + "null";
 } else { 
__p = __p
    + " function() { return (";
__p += ((__t = 
  templates.subschema(Object.assign({
    name: '???',
    required: false,
    schema: schema.additionalProperties,
  }, extras)) ) == null ? "" : __t);
__p = __p
    + "); }";

}

__p = __p
    + ";\n"
    + "\n"
    + "const handlers = Object.create(null);\n"
    + "";

  for (const propName of Object.keys(schema.properties)) {
    const _key = JSON.stringify(propName);
    const _name = propName.match(/^[_a-zA-Z][_a-zA-Z0-9]*$/) ? propName : `[${_key}]`;

__p = __p
    + "handlers[";
__p += ((__t =  JSON.stringify(_name) ) == null ? "" : __t);
__p = __p
    + "] = ";
__p += ((__t = 
  templates.subschema(Object.assign({
    name: _name,
    required: required.indexOf(propName) != -1,
    schema: schema.properties[propName],
  }, extras))
) == null ? "" : __t);

}

__p = __p
    + "\n"
    + "\n"
    + "const objectFunctions = {\n"
    + "  hasOwnProperty(name) { return objectProxyHandler.has(null, name); },\n"
    + "};\n"
    + "\n"
    + "\n"
    + "let defined = false;\n"
    + "const objectProxyHandler = {\n"
    + "  isExtensible() { return false; },\n"
    + "  has(target, prop) {\n"
    + "    return (prop in handlers && handlers[prop].isDefined())\n"
    + "        || (additionalPropertyHandler\n"
    + "            ? (prop in additionalProperties && additionalProperties[prop].isDefined())\n"
    + "            : false);\n"
    + "  },\n"
    + "  get(target, prop) {\n"
    + "    if (prop === symbols.Schema) {\n"
    + "      return objectHandler.schema;\n"
    + "    }\n"
    + "    if (prop in handlers) {\n"
    + "      return handlers[prop].get();\n"
    + "    }\n"
    + "    if (prop in objectFunctions) {\n"
    + "      return objectFunctions[prop];\n"
    + "    }\n"
    + "    return undefined;\n"
    + "  },\n"
    + "  set(target, prop, v) {\n"
    + "    defined = true;\n"
    + "    if (prop in handlers) {\n"
    + "      handlers[prop].set(v);\n"
    + "      return true;\n"
    + "    } else if (additionalPropertyHandler) {\n"
    + "      if (!(prop in additionalProperties)) {\n"
    + "        additionalProperties[prop] = additionalPropertyHandler(prop);\n"
    + "      }\n"
    + "      additionalProperties[prop].set(v);\n"
    + "      return true;\n"
    + "    } else {\n"
    + "      ";
 if (options.ignoreExtraProperties !== true) {
      
__p = __p
    + "throw new exceptions.InvalidPropertyNameException(";
__p += ((__t =  JSON.stringify(path) ) == null ? "" : __t);
__p = __p
    + " + '/' + prop);";

      } else {
        // Just ignore the property.
      
__p = __p
    + "return true;";

      } 
__p = __p
    + "\n"
    + "    }\n"
    + "  },\n"
    + "  deleteProperty(target, prop) {\n"
    + "    if (prop in handlers) {\n"
    + "      handlers[prop].remove();\n"
    + "      return true;\n"
    + "    } else if (additionalPropertyHandler && prop in additionalProperties) {\n"
    + "      delete additionalProperties[prop];\n"
    + "    }\n"
    + "  },\n"
    + "  defineProperty(target, prop, descriptor) {\n"
    + "    return false;\n"
    + "  },\n"
    + "  getOwnPropertyDescriptor(target, prop) {\n"
    + "    if (prop in handlers) {\n"
    + "      return { configurable: true, enumerable: true };\n"
    + "    } else if (additionalPropertyHandler && prop in additionalPropertyHandler) {\n"
    + "      return { configurable: true, enumerable: true };\n"
    + "    }\n"
    + "  },\n"
    + "  ownKeys(target) {\n"
    + "    return [].concat(\n"
    + "      Object.keys(handlers),\n"
    + "      additionalPropertyHandler ? Object.keys(additionalProperties) : []\n"
    + "    );\n"
    + "  },\n"
    + "};\n"
    + "\n"
    + "\n"
    + "const proxy = new Proxy({}, objectProxyHandler);\n"
    + "\n"
    + "const objectHandler = {\n"
    + "  set(v) {\n"
    + "    if (v === undefined) {\n"
    + "      defined = false;\n"
    + "      return;\n"
    + "    }\n"
    + "\n"
    + "    defined = true;\n"
    + "    for (const key of Object.keys(v)) {\n"
    + "      proxy[key] = v[key];\n"
    + "    }\n"
    + "\n"
    + "    // Validate required fields.\n"
    + "    ";
 for (const key of required) { 
__p = __p
    + "\n"
    + "    if (!(";
__p += ((__t =  JSON.stringify(key) ) == null ? "" : __t);
__p = __p
    + " in v)) {\n"
    + "      throw new exceptions.RequiredValueMissingException(";
__p += ((__t =  JSON.stringify(path) ) == null ? "" : __t);
__p = __p
    + " + '/' + ";
__p += ((__t =  JSON.stringify(key) ) == null ? "" : __t);
__p = __p
    + ");\n"
    + "    }";
 } 
__p = __p
    + "\n"
    + "  },\n"
    + "  get() {\n"
    + "    if (";
__p += ((__t =  options.allowAccessUndefinedObjects === true ) == null ? "" : __t);
__p = __p
    + " || this.isDefined()) {\n"
    + "      return proxy;\n"
    + "    } else {\n"
    + "      return ";
__p += ((__t =  'default' in schema ? JSON.stringify(schema.default) : 'undefined' ) == null ? "" : __t);
__p = __p
    + ";\n"
    + "    }\n"
    + "  },\n"
    + "  isDefined() {\n"
    + "    return defined\n"
    + "      && (Object.keys(handlers).some(function(x) { return handlers[x].isDefined(); })\n"
    + "          || Object.keys(additionalProperties).some(function(x) {\n"
    + "            return additionalProperties[x].isDefined();\n"
    + "          }));\n"
    + "  },\n"
    + "  remove() { this.set(undefined); },\n"
    + "  schema: ";
__p += ((__t =  JSON.stringify(schema) ) == null ? "" : __t);
__p = __p
    + ",\n"
    + "};\n"
    + "\n"
    + "return objectHandler;\n"
    + "\n"
    + "})()\n"
    + "";
  };

  return __p;
}
