// @angular/service-worker/worker/cache.js
var ScopedCache = (function () {
    function ScopedCache(delegate, prefix) {
        this.delegate = delegate;
        this.prefix = prefix;
    }
    ScopedCache.prototype.load = function (cache, req) {
        return this.delegate.load(this.prefix + cache, req);
    };
    ScopedCache.prototype.store = function (cache, req, resp) {
        return this.delegate.store(this.prefix + cache, req, resp);
    };
    ScopedCache.prototype.remove = function (cache) {
        return this.delegate.remove(this.prefix + cache);
    };
    ScopedCache.prototype.invalidate = function (cache, req) {
        return this.delegate.invalidate(this.prefix + cache, req);
    };
    ScopedCache.prototype.keys = function () {
        var _this = this;
        return this
            .delegate
            .keys()
            .then(function (keys) { return keys
            .filter(function (key) { return key.startsWith(_this.prefix); })
            .map(function (key) { return key.substr(_this.prefix.length); }); });
    };
    ScopedCache.prototype.keysOf = function (cache) {
        return this
            .delegate
            .keysOf(this.prefix + cache);
    };
    return ScopedCache;
}());
export { ScopedCache };
