// tsutils/src/wrap.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ts = require("typescript");
function flattenAst(sourceFile) {
    return getConvertedAst(sourceFile).flat;
}
exports.flattenAst = flattenAst;
function wrapAst(sourceFile) {
    return getConvertedAst(sourceFile).wrapped;
}
exports.wrapAst = wrapAst;
var CACHE = new WeakMap();
function getConvertedAst(sourceFile) {
    var result = CACHE.get(sourceFile);
    if (result !== undefined)
        return result;
    result = convertAst(sourceFile);
    CACHE.set(sourceFile, result);
    return result;
}
function convertAst(sourceFile) {
    var wrapped = {
        node: sourceFile,
        parent: undefined,
        children: [],
        next: undefined,
        skip: undefined,
    };
    var flat = [];
    var current = wrapped;
    var previous = current;
    ts.forEachChild(sourceFile, function wrap(node) {
        flat.push(node);
        var parent = current;
        previous.next = current = {
            node: node,
            parent: parent,
            children: [],
            next: undefined,
            skip: undefined,
        };
        if (previous !== parent)
            setSkip(previous, current);
        previous = current;
        parent.children.push(current);
        ts.forEachChild(node, wrap);
        current = parent;
    });
    return {
        wrapped: wrapped,
        flat: flat,
    };
}
function setSkip(node, skip) {
    do {
        node.skip = skip;
        node = node.parent;
    } while (node !== skip.parent);
}
